components/browser/state/src/main/java/mozilla/components/browser/state/state/SessionState.kt

```diff
@@ -45,61 +50,157 @@ interface SessionState {
-         * Created to handle a CustomTabs intent
-         */
-        CUSTOM_TAB,
+            /**
+             * Created to handle an ACTION_SEARCH and ACTION_WEB_SEARCH intent.
+             */
+            data class ActionSearch(override val caller: ExternalPackage?) : External(3, caller)
 
-        /**
-         * User interacted with the home screen
-         */
-        HOME_SCREEN,
+            /**
+             * Created to handle a CustomTabs intent of external origin.
+             */
+            data class CustomTab(override val caller: ExternalPackage?) : External(4, caller)
+        }
 
         /**
-         * User interacted with a menu
+         * Describes sessions of internal origin, i.e. from within of the application.
          */
-        MENU,
+        sealed class Internal(id: Int) : Source(id) {
+            /**
+             * User interacted with the home screen.
+             */
+            object HomeScreen : Internal(5)
 
-        /**
-         * User opened a new tab
-         */
-        NEW_TAB,
+            /**
+             * User interacted with a menu.
+             */
+            object Menu : Internal(6)
 
-        /**
-         * Default value and for testing purposes
-         */
-        NONE,
+            /**
+             * User opened a new tab.
+             */
+            object NewTab : Internal(7)
 
-        /**
-         * Default value and for testing purposes
-         */
-        TEXT_SELECTION,
+            /**
+             * Default value and for testing purposes.
+             */
+            object None : Internal(8)
 
-        /**
-         * User entered a URL or search term
-         */
-        USER_ENTERED,
+            /**
+             * Default value and for testing purposes.
+             */
+            object TextSelection : Internal(9)
+
+            /**
+             * User entered a URL or search term.
+             */
+            object UserEntered : Internal(10)
+
+            /**
+             * Created to handle a CustomTabs intent of internal origin.
+             */
+            object CustomTab : Internal(11)
+        }
+    }
+}
+
+/**
+ * Describes a category of an external package.
+ */
+@Suppress("MagicNumber")
+enum class PackageCategory(val id: Int) {
+    UNKNOWN(-1),
+    GAME(0),
+    AUDIO(1),
+    VIDEO(2),
+    IMAGE(3),
+    SOCIAL(4),
+    NEWS(5),
+    MAPS(6),
+    PRODUCTIVITY(7);
 
+    companion object {
         /**
-         * This session was restored
+         * Maps an int category (as it can be obtained from a package manager) to our internal representation.
          */
-        RESTORED
+        fun fromInt(id: Int?): PackageCategory = when (id) {
+            0 -> GAME
+            1 -> AUDIO
+            2 -> VIDEO
+            3 -> IMAGE
+            4 -> SOCIAL
+            5 -> NEWS
+            6 -> MAPS
+            7 -> PRODUCTIVITY
+            -1 -> UNKNOWN
+            null -> UNKNOWN
+            else -> UNKNOWN
```

Comment:
nit: You can access all values of an enum with `values` and could avoid repeating the ids by writing something like:
```Kotlin
fun fromInt(id: Int?): PackageCategory = values().find { category -> category.id == id } ?: UNKNOWN
```